pkg/dns/aws/filewatcher.go

```diff
@@ -0,0 +1,51 @@
+package aws
+
+import (
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/elb"
+	"github.com/aws/aws-sdk-go/service/resourcegroupstaggingapi"
+	"github.com/aws/aws-sdk-go/service/route53"
+)
+
+// StartWatcher starts a file watcher and ensures AWS sessions are using the
+// current ca bundle until a message is received on the stop or error channels.
+func (m *Provider) StartWatcher(operatorReleaseVersion string, stop <-chan struct{}) error {
+	errChan := make(chan error)
+	reloadChan := make(chan bool)
+	go func() {
+		errChan <- m.fileWatcher.Start(stop, reloadChan)
+	}()
+	go func() {
+		errChan <- m.ensureSessionTransport(operatorReleaseVersion, reloadChan)
+	}()
+
+	// Wait for the watcher to exit or an explicit stop.
+	select {
+	case <-stop:
+		return nil
+	case err := <-errChan:
+		return err
+	}
+}
+
+// ensureSessionTransport ensures AWS sessions use the current certificates
+// from the file watcher.
+func (m *Provider) ensureSessionTransport(operatorReleaseVersion string, reloadCh chan bool) error {
+	for {
+		select {
+		case <-reloadCh:
+			sess, err := NewProviderSession(m.config, operatorReleaseVersion, m.fileWatcher.GetFileData())
+			if err != nil {
+				return fmt.Errorf("failed to create dns provider session: %v", err)
+			}
+			m.lock.Lock()
+			m.elb = elb.New(sess, aws.NewConfig().WithRegion(m.config.Region))
+			m.route53 = route53.New(sess)
+			m.tags = resourcegroupstaggingapi.New(sess, aws.NewConfig().WithRegion("us-east-1"))
+			m.lock.Unlock()
```

Comment:
Is this actually thread-safe absent locks near the readers throughout the rest of the code?

Reply:
The `FileWatcher` struct uses a mutex to lock/unlock any mutations to the watched file data. The `handleEvent()` method uses a chan to indicate when the watcher updates the current data of the watched file. The `ensureSessionTransport()` updates the aws dns sessions when a value is received on the chan. `StartWatcher()` runs `ensureSessionTransport()` and filewatcher `Start()`, which blocks until the stop channel is closed and listens for fsnotify events and errors on chans. Let me know if you have any suggestions to improve thread safety.

Reply:
The problem is we have code such as the following:
```go
	outerError := m.tags.GetResourcesPages(&resourcegroupstaggingapi.GetResourcesInput{
		ResourceTypeFilters: []*string{aws.String("route53:hostedzone")},
		TagFilters:          tagFilters,
	}, f)
```
To be thread safe, we now need to take a lock before using `m.tags`:
```go
	m.lock.Lock()
	outerError := m.tags.GetResourcesPages(&resourcegroupstaggingapi.GetResourcesInput{
		ResourceTypeFilters: []*string{aws.String("route53:hostedzone")},
		TagFilters:          tagFilters,
	}, f)
	m.lock.Unlock()
```
And ditto for `m.route53` and `m.elb`.

Reply:
The lock/unlock occurs [here](https://github.com/openshift/cluster-ingress-operator/blob/master/pkg/dns/aws/dns.go#L122-L123) for that tagging api call, [here](https://github.com/openshift/cluster-ingress-operator/blob/master/pkg/dns/aws/dns.go#L266-L267) for the route53 api call, and [here](https://github.com/openshift/cluster-ingress-operator/blob/master/pkg/dns/aws/dns.go#L191-L192) for elb api calls.

I'm getting ready to push a commit that will cause the ingress-operator pod to gracefully restart when the ca bundle changes to simplify the implementation and remove the need to worry about runaway goroutines.